File: sol/1593/1593.py
class Solution:
    def maxUniqueSplit(self, s: str) -> int:
        n = len(s)
        res = 0
        def backtrack(idx, oldSubstrs):
            if idx == n:
                nonlocal res
                res = max(res, len(oldSubstrs))
                return
            for i in range(idx + 1, n + 1):
                if s[idx:i] not in oldSubstrs:
                    oldSubstrs.add(s[idx:i])
                    backtrack(i, oldSubstrs)
                    oldSubstrs.remove(s[idx:i])
        backtrack(0, set())
        return res

File: sol/1593/1593.cpp
#include <unordered_map>

struct Solution {
    void backtrack(int idx, unordered_map<string, bool> &oldSubstrs, int count, string &s, int &res){
        if (idx == s.size()){
            res = max(res, count);
            return;
        }
        for (int i = idx + 1; i <= s.size(); ++i){
            string substr = s.substr(idx, i - idx);
            if (!oldSubstrs[substr]){
                oldSubstrs[substr] = true;
                this->backtrack(i, oldSubstrs, count + 1, s, res);
                oldSubstrs[substr] = false;
            }
        }
    }
    int maxUniqueSplit(string &s) {
        int res = 0;
        unordered_map<string, bool> oldSubstrs;
        this->backtrack(0, oldSubstrs, 0, s, res);
        return res;
    }
};
